### PR TITLE
Spark/ADAM Pipeline

### PR DESCRIPTION
**Pull request by fnothaft**
First copy of working Spark/ADAM pipeline. Some issues remaining, but this is a reasonable working copy for now. Hold off on merging; this depends on changes that are not yet merged into ADAM. I have a custom build with these changes in a [branch of cgl-docker-lib](https://github.com/fnothaft/cgl-docker-lib/tree/adam-599).

### CODE REVIEW

**Comment by heuermh** [comment]
@fnothaft where would you like review comments for this and related?
The diff for https://github.com/bigdatagenomics/adam/compare/master...fnothaft:ssf-plus-599 is quite large.

**Comment by fnothaft** [comment]
@heuermh The ssf-plus-559 is https://github.com/bigdatagenomics/adam/pull/901 and https://github.com/bigdatagenomics/adam/pull/906 as a single branch. Can you comment on all three pull requests on their individual pull request? Thanks!

**Comment by fnothaft** [comment]
@almussel can you review? I think this is good to merge.

**Comment by fnothaft** [comment]
We will merge https://github.com/bigdatagenomics/adam/pull/906 shortly and work on the ADAM issue you reported today, but I think this is ready.

**Comment by almussel** [comment]
Looks good to me!

**Comment by heuermh** [comment]
+1

File: src/toil_scripts/adam_pipeline/spark_toil_script.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+import os
+from subprocess import * #call, check_output
+import sys
+from toil.job import Job
+
+SPARK_MASTER_PORT = "7077"
+HDFS_MASTER_PORT = "8020"
+log = open("python.log", 'a')
+
+# - The tachyon and spark containers need the following ports open:
```
Comment (heuermh) [inline src/toil_scripts/adam_pipeline/spark_toil_script.py]: tachyon &rarr; hadoop here, or remove the comment?

Comment (fnothaft) [inline src/toil_scripts/adam_pipeline/spark_toil_script.py]: I'll remove the comment. We've got the port mappings documented in cgl-docker-lib and are running with `--net=host` anyways.

File: src/toil_scripts/adam_pipeline/spark_toil_script.py [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    """
+    Preprocess inFile with known SNPs snpFile:
+        - mark duplicates
+        - realign indels
+        - recalibrate base quality scores
+    """
+    log.write("adam transform\n")
+    log.flush()
+
+    outFile = ".".join(inFile.split(".")[:-1])+".processed.bam"
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_pipeline/spark_toil_script.py]: Use os.path.splitext() instead?

File: src/toil_scripts/adam_pipeline/spark_toil_script.py [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    for i in range(inputs['numWorkers']):
+        job.addService(WorkerService(masterIP))
+    job.addFollowOnJobFn(download_data, masterIP, inputs)
+
+
+def call_conductor(masterIP, inputs, src, dst):
+    """
+    Invokes the conductor container.
+    """
+    return call(["docker",
```
Comment (hannes-ucsc) [inline src/toil_scripts/adam_pipeline/spark_toil_script.py]: Use `check_call()` and drop the assertions after every invocation of `call_conductor()`

**Comment by heuermh** [comment]
Created a pull request to address review comments here:  https://github.com/fnothaft/toil-scripts/pull/2
@almussel could you review?

**Comment by fnothaft** [comment]
+1

**Comment by fnothaft** [comment]
@hannes-ucsc can you remove the `needs work` tag? I think this is good to go, pending @almussel's review.

@almussel can you review a4f33a4? This mounts ephemeral storage as an HDFS volume. We depend on https://github.com/BD2KGenomics/cgl-docker-lib/pull/68 now.

**Comment by hannes-ucsc** [comment]
Yes, there already are other things in `/mnt/ephemeral` and HDFS should operate within a subdirectory thereof.

I am ok with the workaround of creating `/mnt/ephemeral/hdfs` inside the container for now. In the container it would be `/hdfs/hdfs`, so you may want to consider `-v /mnt/ephemeral/:/ephemeral/:rw` or even `-v /mnt/ephemeral/:/mnt/ephemeral/:rw` to make it less confusing.

In the long run there is `/var/lib/toil` which is bind-mounted by cgcloud from `/mnt/persistent/var/lib/toil`.  Most scripts in toil-scripts set the working dir to that directory (via Toil's `--workDir` flag). You can then use `job.fileStore.getLocalTempDir()` within a job to get a job-specific temporary directory somewhere underneath `--workDir`. The catch is that currently the bind mount is hardwired to the persistent volume, not the ephemeral one which is what you would want. I think that is a mistake and it can easily be changed. In

https://github.com/BD2KGenomics/cgcloud/blob/master/toil/src/cgcloud/toil/toil_box.py#L77

`persistent=True` should be `persistent=False`. I will make that change now. It's up to you if you want to upgrade cgcloud and create a new toil-box image or if you want to go ahead with the hack.

Also think about clean-up.

**Comment by hannes-ucsc** [comment]
@fnothaft, @massie and @heuermh now have write access to the repo which should enable you to add and remove labels.

**Comment by fnothaft** [comment]
> @fnothaft, @massie and @heuermh now have write access to the repo which should enable you to add and remove labels.

Thanks! Worked on my side.

Will update this with the small comments from above tomorrow AM.

**Comment by hannes-ucsc** [comment]
I'm confused. You took off the `needs work` label but you say you will update this again?

**Comment by fnothaft** [comment]
@hannes-ucsc Sorry, may have been a misunderstanding about the tag. I'd assumed it meant something like "this work is incomplete and not ready for review". I'll add the tag back and remove it when all PR comments are addressed.

**Comment by fnothaft** [comment]
Actually, looks like you can't add or remove tags from the GitHub mobile web UI... I'll sort the tags out once I'm back on my work machine.

**Comment by hannes-ucsc** [comment]
OK, let's forget about the label. 

Can this be merged?

**Comment by fnothaft** [comment]
@hannes-ucsc this is good to go from my side now. I made the last fixes to address your earlier comment. This depends on https://github.com/BD2KGenomics/cgl-docker-lib/pull/68, which I saw you just merged.

If you can merge this, I am going to make a follow-on PR to address @almussel's Spark/disk issue that we were discussing via email.

**Pull request merged**